src/contracts/core/DelegationManager.sol

```diff
@@ -785,6 +1018,146 @@ contract DelegationManager is
+        return scaledShares * totalMagnitude / SlashingConstants.PRECISION_FACTOR;
+    }
+
+    /**
+     * @notice staker scaling factor should be initialized and lower bounded to 1e18
+     */
+    function _getStakerScalingFactor(address staker, IStrategy strategy) internal view returns (uint256) {
+        uint256 currStakerScalingFactor = stakerScalingFactors[staker][strategy];
+        if (currStakerScalingFactor == 0) {
+            currStakerScalingFactor = SlashingConstants.PRECISION_FACTOR;
+        }
+        return currStakerScalingFactor;
+    }
+
+    /**
+     * @notice helper to calculate the new staker scaling factor after adding shares. This is only used
+     * when a staker is depositing through the StrategyManager or EigenPodManager. A stakers scaling factor
+     * is only updated when they have new deposits and their shares are being increased.
+     */
+    function _calculateStakerScalingFactor(
+        address staker,
+        IStrategy strategy,
+        uint64 totalMagnitude,
+        uint256 existingShares,
+        uint256 addedShares
+    ) internal view returns (uint256) {
+        uint256 newStakerScalingFactor;
+
+        if (existingShares == 0) {
+            // existing shares are 0, meaning no existing delegated shares. In this case, the new staker scaling factor
+            // is re-initialized to
+            newStakerScalingFactor = SlashingConstants.PRECISION_FACTOR / (totalMagnitude);
+        } else {
+            uint256 currStakerScalingFactor = _getStakerScalingFactor(staker, strategy);
+
+            // TODO: DOUBLE CHECK THIS BEHAVIOR AND OVERFLOWS
+            // staker scaling factor is initialized to PRECISION_FACTOR(1e18) and totalMagnitude is initialized to INITIAL_TOTAL_MAGNITUDE(1e18)
+            // and is monotonically decreasing. You can deduce that the newStakerScalingFactor will never decrease to less than the PRECISION_FACTOR
+            // so this won't round to 0.
+            newStakerScalingFactor = (
+                currStakerScalingFactor * existingShares * totalMagnitude / SlashingConstants.PRECISION_FACTOR
+                    + addedShares * SlashingConstants.PRECISION_FACTOR
+            ) / ((existingShares + addedShares) * totalMagnitude);
+        }
+
+        return newStakerScalingFactor;
+    }
+
+    /**
+     * @notice helper to calculate the withdrawable shares for a staker given their shares and the
+     * current totalMagnitude. `shares` should be the staker's shares in storage in the StrategyManager/EigenPodManager.
+     */
+    function _getWithdrawableShares(
+        address staker,
+        IStrategy strategy,
+        uint256 shares,
+        uint64 currTotalMagnitude
+    ) internal view returns (uint256) {
+        uint256 stakerScalingFactor = _getStakerScalingFactor(staker, strategy);
+        return stakerScalingFactor * shares * currTotalMagnitude / SlashingConstants.PRECISION_FACTOR;
```

Comment:
the magnitude needs to be normalized here? missing another `/ INITIAL_TOTAL_MAGNITUDE`

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/679/commits/a70ed62f13fa8be58c456c8d518ec71e7a8482ab
Changed to `(stakerScalingFactor * shares / PRECISION_FACTOR) * currTotalMagnitude / PRECISION_FACTOR`

Reply:
it should be what's in my message if `INITIAL_TOTAL_MAGNITUDE != PRECISION_FACTOR`